core/store/src/flat_state.rs

```diff
@@ -331,18 +331,56 @@ mod imp {
     }
 }
 
-use borsh::{BorshDeserialize, BorshSerialize};
-
+#[cfg(feature = "protocol_feature_flat_state")]
+use crate::DBCol;
 use crate::{metrics, CryptoHash, Store, StoreUpdate};
+use borsh::{BorshDeserialize, BorshSerialize};
 pub use imp::{FlatState, FlatStateFactory};
 use near_primitives::state::ValueRef;
 use near_primitives::types::{BlockHeight, RawStateChangesWithTrieKey, ShardId};
 use std::collections::{HashMap, HashSet};
+use std::io::Read;
 
-#[derive(BorshSerialize, BorshDeserialize)]
-pub struct KeyForFlatStateDelta {
+#[derive(PartialEq, Eq, Debug)]
+/// Contents of `FlatStateDelta` key in the DB.
+pub struct FlatStateDeltaKey {
+    /// Shard id to which the delta belongs to.
     pub shard_id: ShardId,
+    /// Hash of block on top of which the delta was created.
     pub block_hash: CryptoHash,
+    /// Trie key.
+    pub key: Vec<u8>,
+}
+
+impl FlatStateDeltaKey {
+    #[allow(unused)]
+    /// Should return prefix of all `FlatStateDeltaKey`s corresponding to one flat storage delta.
+    fn db_prefix(shard_id: ShardId, block_hash: CryptoHash) -> Vec<u8> {
+        let mut out = Vec::new();
+        out.extend(shard_id.to_be_bytes());
+        out.extend(block_hash.as_bytes());
```

Comment:
this is part of encoding logic which is duplicated in `encode`, lets extract this to a separate method, something like `fn append_prefix(out: &mut Vec<u8>, shard_id: ShardId, block_hash: CryptoHash)` or `fn encode_prefix(shard_id: ShardId, block_hash: CryptoHash) -> [u8;40]` and then append it